nifi-nar-bundles/nifi-kite-bundle/nifi-kite-processors/src/main/java/org/apache/nifi/processors/kite/InferAvroSchemaFromJSON.java

```diff
@@ -0,0 +1,164 @@
+import org.apache.nifi.annotation.documentation.SeeAlso;
+import org.apache.nifi.annotation.documentation.Tags;
+import org.apache.nifi.components.PropertyDescriptor;
+import org.apache.nifi.flowfile.FlowFile;
+import org.apache.nifi.processor.ProcessContext;
+import org.apache.nifi.processor.ProcessSession;
+import org.apache.nifi.processor.ProcessorInitializationContext;
+import org.apache.nifi.processor.Relationship;
+import org.apache.nifi.processor.exception.ProcessException;
+import org.apache.nifi.processor.io.InputStreamCallback;
+import org.apache.nifi.processor.io.OutputStreamCallback;
+import org.apache.nifi.processor.util.StandardValidators;
+import org.kitesdk.data.spi.JsonUtil;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.Collections;
+import java.util.concurrent.atomic.AtomicReference;
+
+@Tags({"kite", "json", "avro", "infer", "schema"})
+@SeeAlso({InferAvroSchemaFromJSON.class})
+@InputRequirement(InputRequirement.Requirement.INPUT_REQUIRED)
+@CapabilityDescription("Creates an Avro schema from JSON data. The Avro schema is inferred by examining the fields " +
+        "in the JSON input. The Avro schema generated by kite will use the same names present in the incoming JSON payload")
+public class InferAvroSchemaFromJSON
+        extends AbstractKiteProcessor {
+
+    public static final PropertyDescriptor RECORD_NAME = new PropertyDescriptor.Builder()
+            .name("Avro Record Name")
+            .description("Value to be placed in the Avro record schema \"name\" field.")
+            .required(true)
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor NUM_RECORDS_TO_ANALYZE = new PropertyDescriptor.Builder()
+            .name("Number of records to analyze")
+            .description("Number of records that should be analyzed by kite to infer the Avro schema")
+            .required(true)
+            .defaultValue("10")
+            .addValidator(StandardValidators.INTEGER_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor CHARSET = new PropertyDescriptor.Builder()
+            .name("Charset")
+            .description("Character encoding of CSV data.")
+            .required(true)
+            .defaultValue("UTF-8")
+            .addValidator(StandardValidators.NON_EMPTY_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PRETTY_AVRO_OUTPUT = new PropertyDescriptor.Builder()
+            .name("Pretty Avro Output")
+            .description("If true the Avro output will be formatted.")
+            .required(true)
+            .defaultValue("true")
+            .allowableValues("true", "false")
+            .addValidator(StandardValidators.BOOLEAN_VALIDATOR)
```

Comment:
This doesn't hurt anything but generally I would recommend against using a validator in conjunction with allowable values, just because i feel it's cleaner. You can leave it in if you prefer, though.